File: tests/debug/pybind/controllers/debug_integratedbdotcontroller_mc.cpp
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <random>
#include <stdexcept>

#include <Eigen/Dense>

#include <saltro/limits.h>
#include <saltro/optimizer/warm_start.h>
#include <saltro/pybind/satellite.h>

using namespace saltro;

namespace {

constexpr double PI = 3.14159265358979323846;
constexpr double DEG2RAD = PI / 180.0;
constexpr double SEC_PER_CENTURY = 36525.0 * 86400.0;

constexpr int NUM_SAMPLES = 5;
constexpr double DT_SECONDS = 10.0;
constexpr double SIM_SECONDS = 1000.0;
constexpr int N = static_cast<int>(SIM_SECONDS / DT_SECONDS) + 1;
constexpr double TUMBLE_STOP_THRESHOLD = 0.5 * DEG2RAD;
constexpr unsigned SEED = 20260302u;

Eigen::Vector3d randomUnitVector(std::mt19937& rng) {
    std::normal_distribution<double> normal(0.0, 1.0);
    Eigen::Vector3d v(normal(rng), normal(rng), normal(rng));
    const double n = v.norm();
    if (n < 1e-12) {
        return Eigen::Vector3d::UnitX();
    }
    return v / n;
}

Eigen::Matrix3d randomInertia(std::mt19937& rng) {
    std::uniform_real_distribution<double> mass_dist(4.0, 60.0);
    std::uniform_real_distribution<double> size_dist(0.10, 0.55);

    const double mass = mass_dist(rng);
    const double lx = size_dist(rng);
    const double ly = size_dist(rng);
    const double lz = size_dist(rng);

    Eigen::Matrix3d J = Eigen::Matrix3d::Zero();
    J(0, 0) = (mass / 12.0) * (ly * ly + lz * lz);
    J(1, 1) = (mass / 12.0) * (lx * lx + lz * lz);
    J(2, 2) = (mass / 12.0) * (lx * lx + ly * ly);
    return J;
}

void makeRandomSatelliteCase(
    std::mt19937& rng,
    PlannerSettings& settings,
    Eigen::Matrix3d& J,
    std::unique_ptr<Satellite>& satellite
) {
    settings = PlannerSettings();
    settings.init_traj.initcontroller = 2;

    settings.disturbances.plan_for_aero = false;
    settings.disturbances.plan_for_gg = false;
    settings.disturbances.plan_for_srp = false;
    settings.disturbances.plan_for_prop = false;
    settings.disturbances.plan_for_gendist = false;
    settings.disturbances.plan_for_resdipole = false;

    settings.num_passes = 1;
    settings.passes[0].dt = DT_SECONDS;

    J = randomInertia(rng);
    satellite = std::make_unique<Satellite>(J, settings);

    const double Javg = std::max(1e-6, J.trace() / 3.0);

    std::uniform_int_distribution<int> n_mtq_dist(2, 3);
    std::uniform_int_distribution<int> n_rw_dist(1, 3);
    std::uniform_real_distribution<double> scale_dist(0.7, 1.4);

    const int n_mtq = n_mtq_dist(rng);
    const int n_rw = n_rw_dist(rng);

    const double mtq_base = std::clamp(0.8 * std::sqrt(Javg), 0.03, 0.35);
    const double rw_torque_base = std::clamp(0.006 * Javg, 8e-5, 8e-3);
    const double rw_hmax_base = std::clamp(0.35 * Javg, 0.005, 0.12);

    for (int i = 0; i < n_mtq; ++i) {
        const Eigen::Vector3d axis = randomUnitVector(rng);
        const double max_dipole = std::clamp(mtq_base * scale_dist(rng), 0.02, 0.40);
        satellite->addMTQ(axis, max_dipole);
    }

    for (int i = 0; i < n_rw; ++i) {
        const Eigen::Vector3d axis = randomUnitVector(rng);
        const double max_torque = std::clamp(rw_torque_base * scale_dist(rng), 5e-5, 0.010);
        const double rw_J = std::clamp(0.015 * Javg * scale_dist(rng), 5e-6, 2e-3);
        const double h_max = std::clamp(rw_hmax_base * scale_dist(rng), 0.004, 0.16);
        satellite->addRW(axis, max_torque, rw_J, 0.0, h_max);
    }
}

Satellite::VecX randomInitialState(const Satellite& satellite, std::mt19937& rng) {
    Satellite::VecX x0 = Satellite::VecX::Zero(satellite.stateDim());

    std::uniform_real_distribution<double> w_mag_dist(4.0 * DEG2RAD, 15.0 * DEG2RAD);
    const double w_mag = w_mag_dist(rng);
    x0.segment<3>(Satellite::AV_INDEX) = w_mag * randomUnitVector(rng);

    std::uniform_real_distribution<double> angle_dist(-PI, PI);
    const double angle = angle_dist(rng);
    const Eigen::Vector3d axis = randomUnitVector(rng);

    Eigen::Vector4d q;
    q(0) = std::cos(0.5 * angle);
    q.segment<3>(1) = axis * std::sin(0.5 * angle);
    q.normalize();
    x0.segment<4>(Satellite::QUAT_INDEX) = q;

    return x0;
}

void makeConstantEnvironment(
    Eigen::VectorXd& jtime,
    Eigen::MatrixXd& q_goal,
    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ>& R,
    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ>& V,
    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ>& B,
    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ>& S,
    Eigen::Matrix<double, 1, limits::MAX_LENGTH_TRAJ>& rho
) {
    jtime = Eigen::VectorXd::Zero(N);
    q_goal = Eigen::MatrixXd::Zero(4, N);

    R.setZero();
    V.setZero();
    B.setZero();
    S.setZero();
    rho.setZero();

    const Eigen::Vector3d B_const(2.2e-5, -1.6e-5, 3.1e-5);

    for (int k = 0; k < N; ++k) {
        const double t_sec = static_cast<double>(k) * DT_SECONDS;
        jtime(k) = 0.25 + t_sec / SEC_PER_CENTURY;
        q_goal(0, k) = 1.0;

        B.col(k) = B_const;
    }
}

} // namespace

int main() {
    std::mt19937 rng(SEED);

    Eigen::VectorXd jtime;
    Eigen::MatrixXd q_goal;
    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R;
    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> V;
    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> B;
    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> S;
    Eigen::Matrix<double, 1, limits::MAX_LENGTH_TRAJ> rho;

    makeConstantEnvironment(jtime, q_goal, R, V, B, S, rho);

    int n_pass = 0;

    for (int sample = 0; sample < NUM_SAMPLES; ++sample) {
        PlannerSettings settings;
        Eigen::Matrix3d J = Eigen::Matrix3d::Zero();
        std::unique_ptr<Satellite> satellite;
        makeRandomSatelliteCase(rng, settings, J, satellite);

        const Satellite::VecX x0 = randomInitialState(*satellite, rng);

        Eigen::MatrixXd X = Eigen::MatrixXd::Zero(satellite->stateDim(), N);
        Eigen::MatrixXd U = Eigen::MatrixXd::Zero(satellite->controlDim(), N);

        bool ok = false;
        try {
            ok = optimizer::warm_start(
                settings,
                *satellite,
                x0,
                jtime,
                q_goal,
                N,
                R,
                V,
                B,
                S,
                rho,
                X,
                U
            );
        } catch (const std::exception& exc) {
            std::cout
                << "[sample " << std::setw(3) << std::setfill('0') << sample << "] "
                << "warm_start exception: " << exc.what() << " | "
                << "nMTQ=" << satellite->numMTQ() << " nRW=" << satellite->numRW() << " DT=" << DT_SECONDS
                << '\n';
            continue;
        }

        if (!ok) {
            std::cout
                << "[sample " << std::setw(3) << std::setfill('0') << sample << "] "
                << "warm_start failed\n";
            continue;
        }

        const double initial_w = x0.segment<3>(Satellite::AV_INDEX).norm();
        const double final_w = X.col(N - 1).segment<3>(Satellite::AV_INDEX).norm();
        const bool passed = std::isfinite(final_w) && (final_w <= TUMBLE_STOP_THRESHOLD) && (final_w < initial_w);
        n_pass += static_cast<int>(passed);

        std::cout
            << "[sample " << std::setw(3) << std::setfill('0') << sample << "] "
            << (passed ? "PASS" : "FAIL") << ' '
            << "nMTQ=" << satellite->numMTQ() << " nRW=" << satellite->numRW() << ' '
            << "|w0|=" << std::fixed << std::setprecision(6) << initial_w << ' '
            << "|wf|=" << std::fixed << std::setprecision(6) << final_w << ' '
            << "Jdiag=[" << J(0, 0) << ", " << J(1, 1) << ", " << J(2, 2) << "]\n";
    }

    std::cout << "\nMonte Carlo summary: " << n_pass << '/' << NUM_SAMPLES << " passed\n";

    return 0;
}

File: tests/debug/pybind/controllers/debug_integratedbdotcontroller_mc.py
import sys
from pathlib import Path

import matplotlib
try:
	# Prefer Tk on WSL to avoid Qt/xcb plugin issues; require tkinter to be present.
	import tkinter  # type: ignore
	matplotlib.use("TkAgg")
except Exception as exc:
	print(f"[warn] Falling back to Agg backend ({exc})")
	matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np


def _find_repo_root(start: Path) -> Path:
	for parent in [start] + list(start.parents):
		if (parent / "build").exists() and (parent / "tests").exists():
			return parent
	raise RuntimeError("Could not locate repository root from script path")


ROOT = _find_repo_root(Path(__file__).resolve())
sys.path.insert(0, str(ROOT / "build"))

import saltro_py


PI = np.pi
DEG2RAD = PI / 180.0
SEC_PER_CENTURY = 36525.0 * 86400.0


NUM_SAMPLES = 5
DT_SECONDS = 10.0
SIM_SECONDS = 1000.0
N = int(SIM_SECONDS / DT_SECONDS) + 1
TUMBLE_STOP_THRESHOLD = 0.5 * DEG2RAD
SEED = 20260302


def random_unit_vector(rng: np.random.Generator) -> np.ndarray:
	v = rng.normal(size=3)
	n = np.linalg.norm(v)
	if n < 1e-12:
		return np.array([1.0, 0.0, 0.0])
	return v / n


def random_inertia(rng: np.random.Generator) -> np.ndarray:
	mass = rng.uniform(4.0, 60.0)
	lx = rng.uniform(0.10, 0.55)
	ly = rng.uniform(0.10, 0.55)
	lz = rng.uniform(0.10, 0.55)

	J = np.zeros((3, 3))
	J[0, 0] = (mass / 12.0) * (ly * ly + lz * lz)
	J[1, 1] = (mass / 12.0) * (lx * lx + lz * lz)
	J[2, 2] = (mass / 12.0) * (lx * lx + ly * ly)
	return J


def make_random_satellite_case(rng: np.random.Generator):
	settings = saltro_py.PlannerSettings()
	settings.init_traj.initcontroller = 2

	settings.disturbances.plan_for_aero = False
	settings.disturbances.plan_for_gg = False
	settings.disturbances.plan_for_srp = False
	settings.disturbances.plan_for_prop = False
	settings.disturbances.plan_for_gendist = False
	settings.disturbances.plan_for_resdipole = False

	settings.num_passes = 1
	settings.passes[0].dt = DT_SECONDS

	J = random_inertia(rng)
	satellite = saltro_py.Satellite(J, settings)

	Javg = max(1e-6, np.trace(J) / 3.0)
	n_mtq = int(rng.integers(2, 4))
	n_rw = int(rng.integers(1, 4))

	mtq_base = float(np.clip(0.8 * np.sqrt(Javg), 0.03, 0.35))
	rw_torque_base = float(np.clip(0.006 * Javg, 8e-5, 8e-3))
	rw_hmax_base = float(np.clip(0.35 * Javg, 0.005, 0.12))

	for _ in range(n_mtq):
		axis = random_unit_vector(rng)
		max_dipole = float(np.clip(mtq_base * rng.uniform(0.7, 1.4), 0.02, 0.40))
		satellite.addMTQ(axis, max_dipole)

	for _ in range(n_rw):
		axis = random_unit_vector(rng)
		max_torque = float(np.clip(rw_torque_base * rng.uniform(0.7, 1.4), 5e-5, 0.010))
		rw_J = float(np.clip(0.015 * Javg * rng.uniform(0.7, 1.4), 5e-6, 2e-3))
		h_max = float(np.clip(rw_hmax_base * rng.uniform(0.7, 1.4), 0.004, 0.16))
		satellite.addRW(axis, max_torque, rw_J, 0.0, h_max)

	return settings, satellite, J


def random_initial_state(satellite: "saltro_py.Satellite", rng: np.random.Generator) -> np.ndarray:
	x0 = np.zeros(satellite.stateDim)

	w_mag = rng.uniform(4.0 * DEG2RAD, 15.0 * DEG2RAD)
	x0[saltro_py.Satellite.AV_INDEX:saltro_py.Satellite.AV_INDEX + 3] = w_mag * random_unit_vector(rng)

	angle = rng.uniform(-PI, PI)
	axis = random_unit_vector(rng)
	q = np.zeros(4)
	q[0] = np.cos(0.5 * angle)
	q[1:] = axis * np.sin(0.5 * angle)
	q /= np.linalg.norm(q)
	x0[saltro_py.Satellite.QUAT_INDEX:saltro_py.Satellite.QUAT_INDEX + 4] = q

	return x0


def make_constant_environment():
	t_seconds = np.arange(N, dtype=float) * DT_SECONDS
	jtime = 0.25 + t_seconds / SEC_PER_CENTURY

	R = np.zeros((3, N), dtype=float)
	V = np.zeros((3, N), dtype=float)
	S = np.zeros((3, N), dtype=float)
	rho = np.zeros(N, dtype=float)

	B_const = np.array([2.2e-5, -1.6e-5, 3.1e-5], dtype=float)
	B = np.repeat(B_const[:, None], N, axis=1)

	q_goal = np.zeros((4, N), dtype=float)
	q_goal[0, :] = 1.0
	boresight = np.zeros((3, N), dtype=float)
	boresight[0, :] = 1.0

	return t_seconds, jtime, q_goal, boresight, R, V, B, S, rho


def plot_sample(sample_idx, t, X, U, satellite, initial_w, final_w, passed):
	av_idx = saltro_py.Satellite.AV_INDEX
	omega = X[av_idx:av_idx + 3, :]
	omega_norm = np.linalg.norm(omega, axis=0)

	n_mtq = satellite.numMTQ
	n_rw = satellite.numRW

	fig = plt.figure(figsize=(13, 9))
	status = "PASS" if passed else "FAIL"
	fig.suptitle(
		f"MC Sample {sample_idx:03d} | {status} | nMTQ={n_mtq}, nRW={n_rw} | "
		f"|w0|={initial_w:.4f} rad/s, |wf|={final_w:.4f} rad/s",
		fontsize=11,
	)

	ax_w = fig.add_subplot(2, 2, 1)
	ax_w.plot(t, omega[0, :], label="wx")
	ax_w.plot(t, omega[1, :], label="wy")
	ax_w.plot(t, omega[2, :], label="wz")
	ax_w.plot(t, omega_norm, "k--", lw=1.4, label="||w||")
	ax_w.hlines(TUMBLE_STOP_THRESHOLD, t[0], t[-1], colors="r", linestyles=":", linewidth=1.3, label="threshold")
	ax_w.set_title("Angular Rate")
	ax_w.set_xlabel("Time [s]")
	ax_w.set_ylabel("rad/s")
	ax_w.grid(True)
	ax_w.legend(fontsize=8, ncol=2)

	ax_mtq = fig.add_subplot(2, 2, 2)
	for i in range(n_mtq):
		(line,) = ax_mtq.plot(t, U[i, :], label=f"mtq{i}")
		ulim = float(satellite.getMTQ(i).u_max)
		color = line.get_color()
		ax_mtq.hlines(ulim, t[0], t[-1], colors=color, linestyles=":", linewidth=1.1)
		ax_mtq.hlines(-ulim, t[0], t[-1], colors=color, linestyles=":", linewidth=1.1)
	if n_mtq > 0:
		ax_mtq.plot(t, np.linalg.norm(U[:n_mtq, :], axis=0), "k--", lw=1.2, label="||u_mtq||")
	ax_mtq.set_title("MTQ Control")
	ax_mtq.set_xlabel("Time [s]")
	ax_mtq.set_ylabel("dipole")
	ax_mtq.grid(True)
	ax_mtq.legend(fontsize=8, ncol=2)

	ax_rw = fig.add_subplot(2, 2, 3)
	for i in range(n_rw):
		ui = n_mtq + i
		(line,) = ax_rw.plot(t, U[ui, :], label=f"rw{i}")
		ulim = float(satellite.getRW(i).u_max)
		color = line.get_color()
		ax_rw.hlines(ulim, t[0], t[-1], colors=color, linestyles=":", linewidth=1.1)
		ax_rw.hlines(-ulim, t[0], t[-1], colors=color, linestyles=":", linewidth=1.1)
	if n_rw > 0:
		ax_rw.plot(t, np.linalg.norm(U[n_mtq:, :], axis=0), "k--", lw=1.2, label="||u_rw||")
	ax_rw.set_title("RW Control")
	ax_rw.set_xlabel("Time [s]")
	ax_rw.set_ylabel("torque")
	ax_rw.grid(True)
	ax_rw.legend(fontsize=8, ncol=2)

	ax_text = fig.add_subplot(2, 2, 4)
	ax_text.axis("off")
	msg = (
		f"Result: {status}\n"
		f"Samples uses same criterion as C++ test\n"
		f"Pass iff final ||w|| <= {TUMBLE_STOP_THRESHOLD:.6f} rad/s\n"
		f"and final ||w|| < initial ||w||"
	)
	ax_text.text(0.05, 0.65, msg, fontsize=10, va="top")

	plt.tight_layout()


def main():
	rng = np.random.default_rng(SEED)
	t, jtime, q_goal, boresight, R, V, B, S, rho = make_constant_environment()

	n_pass = 0

	for sample in range(NUM_SAMPLES):
		settings, satellite, J = make_random_satellite_case(rng)
		x0 = random_initial_state(satellite, rng)

		try:
			ok, X, U = saltro_py.warm_start(
				settings,
				satellite,
				x0,
				jtime,
				q_goal,
				boresight,
				R,
				V,
				B,
				S,
				rho,
			)
		except RuntimeError as exc:
			print(
				f"[sample {sample:03d}] warm_start exception: {exc} | "
				f"nMTQ={satellite.numMTQ} nRW={satellite.numRW} DT={DT_SECONDS}"
			)
			continue

		if not ok:
			print(f"[sample {sample:03d}] warm_start failed")
			continue

		initial_w = float(np.linalg.norm(x0[saltro_py.Satellite.AV_INDEX:saltro_py.Satellite.AV_INDEX + 3]))
		final_w = float(np.linalg.norm(X[saltro_py.Satellite.AV_INDEX:saltro_py.Satellite.AV_INDEX + 3, -1]))
		passed = bool(np.isfinite(final_w) and final_w <= TUMBLE_STOP_THRESHOLD and final_w < initial_w)
		n_pass += int(passed)

		print(
			f"[sample {sample:03d}] {'PASS' if passed else 'FAIL'} "
			f"nMTQ={satellite.numMTQ} nRW={satellite.numRW} "
			f"|w0|={initial_w:.6f} |wf|={final_w:.6f} "
			f"Jdiag={np.diag(J)}"
		)

		plot_sample(sample, t, X, U, satellite, initial_w, final_w, passed)

	print(f"\nMonte Carlo summary: {n_pass}/{NUM_SAMPLES} passed")
	plt.show()


if __name__ == "__main__":
	main()

